### ISSUE TITLE
Improve URL handlings

### ISSUE THREAD

**Issue by kyasu1** [dillonkearns/elm-pages#389]
This issue is a bit long, but I think the following will solve several problems regarding the current URL handling I have, which I tried to clarify on Discord but could not convince you.

## Query parameters are dropped

**To Reproduce**

If you have *internal* links something like belows and click one of them,

ttps://elm-pages.com/docs?q=Foo
ttps://elm-pages.com/docs?q=Bar

The browser trrigers the hard reload and the query parameters are dropped. This prevents us from implementing a page with query parameters as states, like serching or sorting contents etc.

**To Resolve**

- Include the equality check of `url.query` at the `navigatingToSamePage` condition
- Should not drop query parameters at `BrowserPushUrl`

```diff
diff --git a/src/Pages/Internal/Platform.elm b/src/Pages/Internal/Platform.elm
index 9b2fcc87..66f1add8 100644
--- a/src/Pages/Internal/Platform.elm
+++ b/src/Pages/Internal/Platform.elm
@@ -392,7 +392,7 @@ update config appMsg model =
                     let
                         navigatingToSamePage : Bool
                         navigatingToSamePage =
-                            url.path == model.url.path
+                            url.path == model.url.path && url.query == model.url.query
                     in
                     if navigatingToSamePage then
                         -- this is a workaround for an issue with anchor fragment navigation
@@ -403,7 +403,7 @@ update config appMsg model =
 
                     else
                         ( model
-                        , BrowserPushUrl url.path
+                        , BrowserPushUrl (Url.toString url)
                         )
 
                 Browser.External href ->
```                 

## Unintentional page reload if you click twice on an url without hash 

**To Reproduce**

Visit 

https://elm-pages.com/docs

then click on the `Docs` link on the navigation bar again will cause page reload

**To Resolve**

Do not call reload pages when there is no hash on the url.

```diff
diff --git a/src/Pages/Internal/Platform.elm b/src/Pages/Internal/Platform.elm
index 9b2fcc87..9f3052de 100644
--- a/src/Pages/Internal/Platform.elm
+++ b/src/Pages/Internal/Platform.elm
@@ -390,9 +390,16 @@ update config appMsg model =
             case urlRequest of
                 Browser.Internal url ->
                     let
                         navigatingToSamePage : Bool
                         navigatingToSamePage =
-                            url.path == model.url.path && url.query == model.url.query
+                            url.path == model.url.path && url.query == model.url.query && url.fragment /= Nothing 
                     in
                     if navigatingToSamePage then
                         -- this is a workaround for an issue with anchor fragment navigation
```                         

## Nothing happen when navigating to a url without hash on the same page

The above patches are almost complete, but if we visit the next url

https://elm-pages.com/docs/what-is-elm-pages#pre-rendered-routes

then click on the top-level link without hash

https://elm-pages.com/docs/what-is-elm-pages/

nothing will happen. This should bring the page to the top.

**To Resolve**

Add `ScrollToTop` effect if there is not fragment on `UrlChanged`.

```diff
@@ -425,12 +432,21 @@ update config appMsg model =
 
                 Nothing ->
                     if model.url.path == url.path && model.url.query == url.query then
-                        ( { model
-                            | -- update the URL in case query params or fragment changed
-                              url = url
-                          }
-                        , NoEffect
-                        )
+                        if url.fragment == Nothing then
+                            ( { model
+                                | -- update the URL in case query params or fragment changed
+                                  url = url
+                              }
+                            , ScrollToTop
+                            )
+
+                        else
+                            ( { model
+                                | -- update the URL in case query params or fragment changed
+                                  url = url
+                              }
+                            , NoEffect
+                            )
 
                     else
                         ( model
```

Perhaps I am making a big mistake, but I would be glad if you would consider these issues.

**Comment by dillonkearns** [comment]
Hello @kyasu1, thanks so much for the report and the detailed writeup! This is very helpful.

I added some cypress test cases and made the changes you suggested, and things appear to be working smoothly.

Query parameters are dropped: 84f5982ac705ade2244f794085e1ed98f57530f1

Unintentional page reload if you click twice on an url without hash: cf8f99a57dc2a4c1e38cb7d5f24cc07c8602b203

Nothing happen when navigating to a url without hash on the same page: 966b8b300d67b677da5609b9d93459a1d623bbcd

The test suite is passing with these changes, so I think this is good to go.

**Comment by dillonkearns** [comment]
This is now shipped in Elm package version 10.0.2. I'll close this issue now as I believe it is resolved, but please do try it out and let me know how it goes! And thanks again for the great feedback and suggestions 🙏
